runtime/serde/serde-cbor/common/src/aws/smithy/kotlin/runtime/serde/cbor/CborDeserializer.kt

```diff
@@ -0,0 +1,225 @@
+    override fun deserializeShort(): Short = deserializeNumber { it.toShort() }
+    override fun deserializeLong(): Long = deserializeNumber { it.toLong() }
+
+    private inline fun <reified T : Number> deserializeFloatingPoint(cast: (Number) -> T): T {
+        val number = when (peekMinorByte(buffer)) {
+            Minor.FLOAT16.value -> Cbor.Encoding.Float16.decode(buffer).value
+            Minor.FLOAT32.value -> Cbor.Encoding.Float32.decode(buffer).value
+            Minor.FLOAT64.value -> Cbor.Encoding.Float64.decode(buffer).value
+            else -> {
+                when (T::class) {
+                    Float::class -> Float.fromBits(decodeArgument(buffer).toInt())
+                    Double::class -> Double.fromBits(decodeArgument(buffer).toLong())
+                    else -> throw DeserializationException("Unsupported floating point type: ${T::class}")
+                }
+            }
+        }
+        return cast(number)
+    }
+
+    override fun deserializeFloat(): Float = deserializeFloatingPoint { it.toFloat() }
+    override fun deserializeDouble(): Double = deserializeFloatingPoint { it.toDouble() }
+
+    override fun deserializeBigInteger(): BigInteger = when (val tagId = peekTag(buffer).id) {
+        2uL -> Cbor.Encoding.BigNum.decode(buffer).value
+        3uL -> Cbor.Encoding.NegBigNum.decode(buffer).value
+        else -> throw DeserializationException("Expected tag 2 or 3 for CBOR BigNum, got $tagId")
+    }
+
+    override fun deserializeBigDecimal(): BigDecimal = Cbor.Encoding.DecimalFraction.decode(buffer).value
+
+    override fun deserializeString(): String = Cbor.Encoding.String.decode(buffer).value
+
+    override fun deserializeBoolean(): Boolean = Cbor.Encoding.Boolean.decode(buffer).value
+
+    override fun deserializeDocument(): Document = throw DeserializationException("Document is not a supported CBOR type.")
+
+    override fun deserializeNull(): Nothing? {
+        Cbor.Encoding.Null.decode(buffer)
+        return null
+    }
+
+    override fun deserializeByteArray(): ByteArray = Cbor.Encoding.ByteString.decode(buffer).value
+
+    override fun deserializeInstant(format: TimestampFormat): Instant = Cbor.Encoding.Timestamp.decode(buffer).value
+}
+
+/**
+ * Element iterator used for deserializing lists
+ */
+private class CborElementIterator(
+    val buffer: SdkBufferedSource,
+    val expectedLength: ULong? = null,
+) : Deserializer.ElementIterator,
+    PrimitiveDeserializer by CborPrimitiveDeserializer(buffer) {
+    var currentLength = 0uL
+
+    override fun hasNextElement(): Boolean {
+        if (expectedLength != null) {
+            if (currentLength != expectedLength) {
+                check(!buffer.exhausted()) { "Buffer is unexpectedly exhausted, read $currentLength elements, expected $expectedLength." }
+                currentLength += 1u
+                return true
+            } else {
+                return false
+            }
+        } else {
```

Comment:
We have two functions to hook into here, `hasNextElement` and `nextHasValue`. Do you think incrementing `currentLength` in `nextHasValue` is better? Or is there some other solution?

Originally I tried incrementing `currentLength` in the primitive deserializer (deserializeByte, deserializeString, etc.) but this failed when deserializing lists with non-primitive members (i.e `List<FooStructure>`) because they have their own deserializer functions and don't go through the primitive deserializer.